### ISSUE TITLE
implement UDP equivalent of TcpStream

### ISSUE THREAD

**Issue by decathorpe** [decathorpe/mitmproxy_wireguard#7]
mitmproxy has its own `asyncio.StreamReader/StreamWriter` UDP analog wrapper implementation based on Python's `asyncio` module.

To integrate the UDP functionality from `mitmproxy_wireguard` into mitmproxy, a similar wrapper interface (`UdpStream`?) around the existing UDP functionality will be required.

**Comment by meitinger** [comment]
Since we already have `DatagramReader` and `DatagramWriter`, which basically are UDP equivalents to `StreamReader` and `StreamWriter`, I'm not so sure what `UdpStream` should do (differently)? (But it's very likely that I'm getting the question wrong.)

As I've mentioned in [#5556](https://github.com/mitmproxy/mitmproxy/pull/5556), I think we should consider adding an interface for transparent UDP in general. We can use the current transparent UDP code for inspiration on how to present data "asyncio-compatible", but use a pipe to an external module for data and setup.

More concrete: An abstract subclass of `ServerInstance`, that on `start` will spawn the transparent module, asking for sudo/UAC permission if necessary (ideally abstract property `needs_elevation`) and use a pipe for communication with that module (either whole packets or from:to:data). Maybe abstract `connect_service`, which will use named pipes or unix socket based on the platform. The service (external module) itself will would setup WinDivert on Windows, netfilter on Linux, PF on Mac, with wireguard I'm guessing establishing a connection to the service.

That way all connection handling will be in mitmproxy, and the external module will be responsible solely for transport. (And we don't need to elevate the entire mitmproxy process on *nix.)

**Comment by decathorpe** [comment]
> Since we already have `DatagramReader` and `DatagramWriter`, which basically are UDP equivalents to `StreamReader` and `StreamWriter`, I'm not so sure what `UdpStream` should do (differently)? (But it's very likely that I'm getting the question wrong.)

`UdpStream` uses the virtual network device provided by `mitmproxy_wireguard` to pipe things over a WireGuard tunnel instead of using Python's `asyncio` module and the host machine's native networking stack, so I'd say it's pretty different :)

**Comment by meitinger** [comment]
> `UdpStream` uses the virtual network device provided by `mitmproxy_wireguard` to pipe things over a WireGuard tunnel instead of using Python's `asyncio` module and the host machine's native networking stack, so I'd say it's pretty different :)

I've just had a look at https://github.com/mitmproxy/mitmproxy/pull/5562, yes, `UdpStream` seems the way to go :)
Although maybe only "export" a `WireguardStream` base class/interface, since the connection handling is in the external module anyway? If `WireguardStream` contains the transport protocol, remote and local address, you should have everything for the `ServerManager` connection id.

**Comment by mhils** [comment]
@meitinger: Let me rephrase the question: Do we need a UdpStream implementation in Rust (https://github.com/decathorpe/mitmproxy_wireguard/pull/8), or do we want the Rust parts to just provide a callback which delivers us packet data? I think we want the latter and then just reuse https://github.com/mitmproxy/mitmproxy/blob/1706a9b9fe8c8157824df22634d97c3d695a495d/mitmproxy/proxy/mode_servers.py#L232, or am I missing things?

**Comment by decathorpe** [comment]
Right now the implementation for UDP matches the one for TCP and is pretty simple ... not sure if making it *even simpler* is worthwhile, given that the `UdpStream.read/write` are basically just the callbacks that you're talking about anyway 😅

**Comment by meitinger** [comment]
> @meitinger: I think we want the latter and then just reuse https://github.com/mitmproxy/mitmproxy/blob/1706a9b9fe8c8157824df22634d97c3d695a495d/mitmproxy/proxy/mode_servers.py#L232, or am I missing things?

If I understand @decathorpe correctly, we actually don't need any `handle_udp_datagram`, because matching packets to connections is already done in `UdpStream`. Since https://github.com/mitmproxy/mitmproxy/pull/5556 we're no longer doing anything special with the connection ID, so to me that actually sounds good.

> Right now the implementation for UDP matches the one for TCP and is pretty simple ... not sure if making it _even simpler_ is worthwhile, given that the `UdpStream.read/write` are basically just the callbacks that you're talking about anyway 😅

By using a common Wireguard stream base class, you wouldn't need to specify TCP _and_ UDP in places like
https://github.com/mitmproxy/mitmproxy/blob/2431d6626ed099012b4506bd05bd3d8261dd2f73/mitmproxy/proxy/server.py#L79
And you can unify `handle_*_connection`s, so in https://github.com/mitmproxy/mitmproxy/blob/2431d6626ed099012b4506bd05bd3d8261dd2f73/mitmproxy/proxy/mode_servers.py#L437 you'd simply do `connection_id = (stream.transport_protocol, stream.get_extra_info("peername"), stream.get_extra_info("sockname"))` and it will work for both.

**Comment by decathorpe** [comment]
> > @meitinger: I think we want the latter and then just reuse https://github.com/mitmproxy/mitmproxy/blob/1706a9b9fe8c8157824df22634d97c3d695a495d/mitmproxy/proxy/mode_servers.py#L232, or am I missing things?
> 
> If I understand @decathorpe correctly, we actually don't need any `handle_udp_datagram`, because matching packets to connections is already done in `UdpStream`. Since [mitmproxy/mitmproxy#5556](https://github.com/mitmproxy/mitmproxy/pull/5556) we're no longer doing anything special with the connection ID, so to me that actually sounds good.

Matching packets to connections is currently very rudimentary: For each new received UDP packet, a new stream is created, but responses to that initial packet are sent over the same "connection" (which caches src_addr and dst_addr). I could add persistent "fake" connections that feed new incoming packets existing UdpStreams, probably by keeping a hashtable that maps (src_addr, dst_addr) to UdpStream instances, but doing that means that I'd also need to implement logic for "closing" those fake connections somehow, otherwise they'd leak and accumulate in memory over time.

> > Right now the implementation for UDP matches the one for TCP and is pretty simple ... not sure if making it _even simpler_ is worthwhile, given that the `UdpStream.read/write` are basically just the callbacks that you're talking about anyway sweat_smile
> 
> By using a common Wireguard stream base class, you wouldn't need to specify TCP _and_ UDP in places like https://github.com/mitmproxy/mitmproxy/blob/2431d6626ed099012b4506bd05bd3d8261dd2f73/mitmproxy/proxy/server.py#L79 And you can unify `handle_*_connection`s, so in https://github.com/mitmproxy/mitmproxy/blob/2431d6626ed099012b4506bd05bd3d8261dd2f73/mitmproxy/proxy/mode_servers.py#L437 you'd simply do `connection_id = (stream.transport_protocol, stream.get_extra_info("peername"), stream.get_extra_info("sockname"))` and it will work for both.

That might work (if each stream stored the protocol), but it would make the implementation much more complicated, since all methods on the stream need to do different things depending on if the underlying protocol is TCP or UDP. So moving everything into one class doesn't really make sense to me, since it will not de-duplicate code but only make the implementation more complicated.

**Comment by meitinger** [comment]
> Matching packets to connections is currently very rudimentary: For each new received UDP packet, a new stream is created, but responses to that initial packet are sent over the same "connection" (which caches src_addr and dst_addr).

Oh, in that case I agree with @mhils, then you're better off doing something similar to `handle_udp_datagram` if you don't want to implement matching yourself.

> So moving everything into one class doesn't really make sense to me, since it will not de-duplicate code but only make the implementation more complicated.

I was thinking in the lines of:
```python
class WireguardStream(metaclass=ABCMeta):
    @property
    @abstractmethod
    def transport_protocol(self) -> Literal["tcp","udp"]:
        pass

    @abstractmethod
    async def read(self, n: int) -> bytes:
        pass

    ... # signatures of all other StreamReader/StreamWriter methods required by mitmproxy

class TcpStream(WireguardStream):
    transport_protocol = "tcp"
    # your existing TcpStream

class UdpStream(WireguardStream):
    transport_protocol = "udp"
    # your existing UdpStream
```

**Comment by decathorpe** [comment]
> I was thinking in the lines of

Yeah, that would work - except that I can't do that, because there's no way to define abstract base classes in PyO3 ...

I could make mitmproxy_wireguard a mixed Python/Rust project, but that would mean the build process would become much more manual and even more annoying, which isn't something I'd like to do given that building binary wheels is already a PITA.

**Comment by meitinger** [comment]
> Yeah, that would work - except that I can't do that, because there's no way to define abstract base classes in PyO3 ...

Oh sorry, I didn't know that. In that case you could just define an alias
```python
WireguardStream = Union[TcpStream, UdpStream]
```
in the mitmproxy part of the code.

**Comment by decathorpe** [comment]
Yeah, adding such an alias looks easy enough. 😅

Should I still add a transport_protocol property to both of them?

---

PS: Looks like the CPython C API doesn't support defining abstract base classes in general (not only not supported by PyO3), since those aren't a thing at the C level at all, but completely faked on the Python level with decorators etc. TIL.

**Comment by meitinger** [comment]
> Should I still add a transport_protocol property to both of them?

If `UdpStream` manages connection association, I'd say yes. If not, I don't think the two "streams" can be treated the same, so you'll still need `handle_udp_datagram` and `handle_tcp_connection`, which makes a `transport_protocol` property unnecessary.

**Comment by decathorpe** [comment]
Ok, that's what I thought, thanks for confirming. 👍🏼

**Comment by mhils** [comment]
I think you're slightly on the wrong track here. Let me try to rephrase:

1. For TCP we already have a well-working TcpStream implementation in mitmproxy-wireguard. Nothing to change here.
2. For UDP we can fully reuse the existing UdpReader/UdpWriter implementations in mitmproxy.udp as well as the existing `handle_udp_datagram`. We only need to provide some mock for `DatagramWriter.transport` so that packets can be sent.

The existing Rust-Python API is good as is (one callback for TCP streams, one for individual UDP packets). Does that make more sense now?

**Comment by decathorpe** [comment]
> Does that make more sense now?

Not really 😓

Looking at `mitmproxy.net.udp.DatagramWriter`, the APIs that are called on the `DatagramWriter._transport` look like those implemented by streams (i.e. `send`, `get_extra_info`, `close`).

You're suggesting to write write a `DatagramTransport`-compatible shim around the previous API instead of the new fake-stream-based API?

It might work to adapt `DatagramReader` and `DatagramWriter` to work with both `asyncio.DatagramTransport` and `UdpStream` (at least easier than with the old non-stream-based API) ...

**Comment by mhils** [comment]
Yes, exactly. Basically provide a replacement for DatagramTransport. More details when I'm not on mobile :)

**Comment by mhils** [comment]
Expanding on my previous comment: 

> It might work to adapt `DatagramReader` and `DatagramWriter` to work with both `asyncio.DatagramTransport` and `UdpStream` (at least easier than with the old non-stream-based API) ...

Basically that, but let's not call this thing `UdpStream` (it's not a stream). I'm happy with `RustDatagramTransport` or whatever else you come up with. :)

One problem with the implementation in #8 is that it differs quite a bit from our DatagramReader/DatagramWriter semantics that we already have. The existing mitmproxy implementation provides a reader that can be awated for additional packets whereas #8 just hands over one packet and then returns `b""`. Fixing this would add quite a lot of complexity on the Rust side, I feel it would be easier to provide a replacement for `DatagramTransport` instead. Looking at `mitmproxy/net/udp.py`, our DatagramTransport replacement only needs to implement:

 - `sendto`
 - `get_extra_info`
 - `get_protocol` -> `wait_closed` and `drain`

First, I don't think we call `get_extra_info` for anything (so that could just raise a `NotImplementedError` for now). 
Second, `get_protocol` can either return a dummy object which provides `wait_closed`/`drain` methods that do nothing,  or it just returns None itself (and we start handling that case in `udp.DatagramWriter`). 
Finally, `sendto` needs to be implemented, but that should be rather straightforward. 🙃 

Does that make more sense now? Happy to expand otherwise :)

**Comment by decathorpe** [comment]
> One problem with the implementation in #8 is that it differs quite a bit from our DatagramReader/DatagramWriter semantics that we already have. The existing mitmproxy implementation provides a reader that can be awated for additional packets whereas #8 just hands over one packet and then returns `b""`.

Yeah, my first draft (which I didn't commit) used internal state to associate incoming UDP packets to existing streams, but that would've been too complicated.

> Fixing this would add quite a lot of complexity on the Rust side, I feel it would be easier to provide a replacement for `DatagramTransport` instead. Looking at `mitmproxy/net/udp.py`, our DatagramTransport replacement only needs to implement:
> 
>     * `sendto`
>     * `get_extra_info`
>     * `get_protocol` -> `wait_closed` and `drain`
> 
> 
> First, I don't think we call `get_extra_info` for anything (so that could just raise a `NotImplementedError` for now). Second, `get_protocol` can either return a dummy object which provides `wait_closed`/`drain` methods that do nothing, or it just returns None itself (and we start handling that case in `udp.DatagramWriter`). Finally, `sendto` needs to be implemented, but that should be rather straightforward.
> 
> Does that make more sense now? Happy to expand otherwise :)

That makes sense, thanks for the clarifications.
